File: src/ertmp/modex.c
/**
 * modex.c — Enhanced RTMP v2 ModEx extension mechanism
 *
 * Per E-RTMP v2 §16, ModEx is a signalling mechanism that lets packets
 * carry modifiers or extensions. The two defined types are:
 *   NOP(0)       — placeholder, no-op
 *   TIMESTAMP(1) — nanosecond-precision timestamp offset
 *
 * Serialised structure:
 *   marker(1, 0x80 | type) +
 *   For TIMESTAMP: 8-byte offset (nanoseconds, big-endian UI64)
 *   For NOP: no additional bytes
 */

#include "ertmp.h"
#include <string.h>

#define LRTMP2_MODEX_MARKER 0x80

int lrtmp2_ertmp_modex_parse(lrtmp2_modex_t *modex, const uint8_t *data, size_t len)
{
    if (!modex || !data || len < 1) return LRTMP2_ERR_IO;

    memset(modex, 0, sizeof(*modex));

    uint8_t marker = data[0];
    if ((marker & 0x80) == 0) return LRTMP2_ERR_PROTOCOL; /* missing marker high bit */

    uint8_t type = marker & 0x7F;
    switch (type) {
        case LRTMP2_MODEX_TYPE_NOP:
            modex->type = LRTMP2_MODEX_TYPE_NOP;
            modex->offset = 0;
            return LRTMP2_OK;

        case LRTMP2_MODEX_TYPE_TIMESTAMP:
            if (len < 9) return LRTMP2_ERR_IO;
            modex->type = LRTMP2_MODEX_TYPE_TIMESTAMP;
            for (int i = 0; i < 8; i++) {
                modex->offset = (modex->offset << 8) | data[1 + i];
            }
            return LRTMP2_OK;

        default:
            /* Unknown ModEx type — ignore per §16 graceful-degradation rule */
            modex->type = LRTMP2_MODEX_TYPE_NOP;
            return LRTMP2_OK;
    }
}

size_t lrtmp2_ertmp_modex_write(const lrtmp2_modex_t *modex, uint8_t *buf, size_t buf_size)
{
    if (!modex || !buf || buf_size < 1) return 0;

    buf[0] = (uint8_t)(LRTMP2_MODEX_MARKER | (uint8_t)modex->type);

    switch (modex->type) {
        case LRTMP2_MODEX_TYPE_NOP:
            return 1;

        case LRTMP2_MODEX_TYPE_TIMESTAMP:
            if (buf_size < 9) return 0;
            {
                uint64_t tmp = modex->offset;
                for (int i = 7; i >= 0; i--) {
                    buf[1 + i] = (uint8_t)(tmp & 0xFF);
                    tmp >>= 8;
                }
            }
            return 9;

        default:
            return 1;
    }
}

File: src/ertmp/modex.rs
//! E-RTMP v2 ModEx extension mechanism
//!
//! Mirrors `src/ertmp/modex.c`.

use crate::types::{Modex, ModexType, Result, ErrorCode};

const MODEX_MARKER: u8 = 0x80;

/// Parse a ModEx extension.
pub fn modex_parse(modex: &mut Modex, data: &[u8]) -> Result<()> {
    if data.is_empty() {
        return Err(ErrorCode::Io);
    }

    let marker = data[0];
    if marker & 0x80 == 0 {
        return Err(ErrorCode::Protocol);
    }

    let ty = marker & 0x7F;
    match ty {
        0 => {
            modex.modex_type = ModexType::Nop;
            modex.offset = 0;
        }
        1 => {
            if data.len() < 9 {
                return Err(ErrorCode::Io);
            }
            modex.modex_type = ModexType::Timestamp;
            modex.offset = 0;
            for i in 0..8 {
                modex.offset = (modex.offset << 8) | data[1 + i] as u64;
            }
        }
        _ => {
            // Unknown ModEx type — ignore per §16 graceful-degradation rule
            modex.modex_type = ModexType::Nop;
        }
    }

    Ok(())
}

/// Write a ModEx extension. Returns bytes written.
pub fn modex_write(modex: &Modex, buf: &mut [u8]) -> usize {
    if buf.is_empty() {
        return 0;
    }

    buf[0] = MODEX_MARKER | modex.modex_type as u8;

    match modex.modex_type {
        ModexType::Nop => 1,
        ModexType::Timestamp => {
            if buf.len() < 9 {
                return 0;
            }
            let mut tmp = modex.offset;
            for i in (0..8).rev() {
                buf[1 + i] = (tmp & 0xFF) as u8;
                tmp >>= 8;
            }
            9
        }
    }
}
